_posts/news/2016-12-21-announcing-scalajs-0.6.14.md

```diff
@@ -0,0 +1,134 @@
+This makes it tedious to create an instance of `JQueryAjaxSettings`, since the Scala type system forces us to fill them in, although the fields are optional in JavaScript.
+Previous workarounds include a) using `js.Dynamic.literal(...).asInstanceOf[JQueryAjaxSettings]` or b) using [`JSOptionBuilder`](https://github.com/jducoeur/jsext#jsoptionbuilder).
+
+Scala.js 0.6.14 finally provides language support for this kind of API.
+It is now allowed to have concrete members in `@ScalaJSDefined` traits, *if* their right-hand-side is `= js.undefined` (this implies that their type must be a supertype of `js.UndefOr[Nothing]`).
+With this relaxed rule, we can define `JQueryAjaxSettings` as follows:
+
+{% highlight scala %}
+@ScalaJSDefined
+trait JQueryAjaxSettings extends js.Object {
+  val data: js.UndefOr[js.Object | String | js.Array[Any]] = js.undefined
+  val contentType: js.UndefOr[Boolean | String] = js.undefined
+  val crossDomain: js.UndefOr[Boolean] = js.undefined
+  val success: js.UndefOr[js.Function3[Any, String, JQueryXHR, _]] = js.undefined
+  ...
+}
+{% endhighlight %}
+
+When calling `ajax()`, we can now give an anonymous object that overrides only the `val`s we care about:
+
+{% highlight scala %}
+jQuery.ajax(someURL, new JQueryAjaxSettings {
+  override val crossDomain: js.UndefOr[Boolean] = true
+  override val success: js.UndefOr[js.Function3[Any, String, JQueryXHR, _]] = {
+    js.defined { (data: Any, textStatus: String, xhr: JQueryXHR) =>
+      println("Status: " + textStatus)
+    }
+  }
+})
+{% endhighlight %}
+
+Note that for functions, we use `js.defined { ... }` to drive Scala's type inference.
+Otherwise, it needs to apply two implicit conversions, which is not allowed.
+
+The explicit types are still quite annoying, but they are only necessary in Scala 2.10 and 2.11.
+If you use Scala 2.12, you can omit all the type annotations (but keep `js.defined`), thanks to improved type inference for `val`s and SAM conversions:
+
+{% highlight scala %}
+jQuery.ajax(someURL, new JQueryAjaxSettings {
+  override val crossDomain = true
+  override val success = js.defined { (data, textStatus, xhr) =>
+    println("Status: " + textStatus)
+  }
+})
+{% endhighlight %}
+
+## Export top-level functions
+
+In Scala.js 0.6.13 and earlier, the only way to export something resembling a top-level function was to export the method and its enclosing object, as follows:
+
+{% highlight scala %}
+@JSExport
+object Main {
+  @JSExport
+  def main(): Unit = { ... }
+}
+{% endhighlight %}
+
+However, this requires to call it as `Main().main()` from JavaScript, which shows that it is not actually a top-level function.
+
+In Scala.js 0.6.14, you can use the `@JSExportTopLevel` annotation instead:
+
+{% highlight scala %}
+object Main {
+  @JSExportTopLevel("main")
+  def main(): Unit = { ... }
+}
+{% endhighlight %}
+
+which allows to call `main()` from JavaScript.
+`@JSExportTopLevel` methods must be declared in top-level `object`s.
```

Comment:
Are you deliberately saying top-level for simplicity here? `@JSExportTopLevel` is allowed in any static object.